rules/do-you-have-separate-development-testing-and-production-environments/rule.md

```diff
@@ -1,80 +1,83 @@
   - apply-tags-to-your-azure-resource-groups
   - do-you-identify-development-test-and-production-crm-web-servers-by-colors
   - do-you-always-rename-staging-url-on-azure
 redirects: []
-
+created: 2009-03-10T07:02:13.000Z
+archivedreason: null
+guid: 616246d0-1675-4c1c-b4b0-d4352fe818e1
 ---
+It is important to separate production and non-production environments. Some companies skip the non-production environments because it can be a hassle to copy new files, register DLLs and deploy backend changes. This will usually result in higher support costs and unhappy users due to simple bugs that could have been found in testing.  
 
-It is important to maintain three separate environments for development, testing and production. Some companies skip the testing server because it can be a hassle to copy new files, register DLLs and deploy backend changes. This will usually result in higher support costs and unhappy users due to simple bugs that could have being found in testing.  
 <!--endintro-->
 
-::: greybox
-The old solution is to use build scripts (.bat and .vbs files) to automatically create a setup package that can be used to deploy to testing and production environments. For backend changes, you can either include the change scripts with the setup package (if it's a localised database), or run those scripts as part of your deployment process.
+## What is each environment for?
 
-Read more about setup packages at [SSW's Wise Standard for Products.](http://www.ssw.com.au/ssw/Standards/wisesetup/WiseStandards.aspx)
-:::
+* **Production**: Real data being used by real customers
+* **Staging**: 'Production-like' data.  Used for testing and verification before deploying to Production.  Should have 'production-like' access (i.e. avoid giving devs admin rights).  Should generally have the same specs as production (especially during performance testing).  However, this is not always the case due to cost implications
+* **Test**: Internal environment used by the development team.  Hosted in the same place as the Staging & Production environments.  Access usually not provided to the client
+* **Ephemeral**: Special environment that is spun up on demand for testing of a branch
+* **Development**: Developer environment running on their local machine.  Also, sometimes called 'Local'
 
-::: bad
-Bad Example - using scripts to deploy your product to different environments
-:::
+![Figure: Environment flow](ssw-rules-environments.jpg "Environment flow")
 
-::: greybox
-The best way to manage your product across different environments is to use DevOps tools (like Azure DevOps, GitHub Actions, and others) and Infrastructure as Code (IaC).
-    
-For more information, see our rule: [Do you know how to create Azure resources?](/azure-resources-creating).
-:::
+### Large or Multi-Team Projects
 
-::: good
-Good Example - use DevOps tools and IaC to deploy to different environments
-:::
+For large projects it's recommended to run 3 hosted environments + 1 local:
```

Comment:
Most of the projects over the years have stuck to the following designations:

Hosted: 
* Production (prod)
* Staging (stage)/(staging)
* Devlopment (dev)

Local:
* Local (localhost)

I'd recommend sticking to the list above to be consistent with our real-life projects.

There has been some projects with QA, UA, Test, SIT, PP, PreProd environment designations. Perhaps your rule can include a statement that starting with the simple set of environments is good, and then there can be more environments added when needed for specific processes.